File: crivoDeEratostenesComParalelismo/threads/CrivoParaleloDados.java
import java.math.BigInteger;
import java.util.HashSet;
import java.util.Scanner;
import java.util.Set;

/**
 * CrivoSequencial
 */
public class CrivoParaleloDados extends Thread{

    static private boolean marcados[];
    
    private int inicio;
    private int fim;

    static private Integer divisor = 2;

    CrivoParaleloDados(int indice, int intervalo){
        this.setPosicao(indice,intervalo);
    }

    public void run() {				
		this.filtraMultiplos();
	}

    static private boolean estaMarcado(int numero){
        return (marcados[numero]);
    }

    private void marcaNaoPrimo(int multiplo){
        marcados[multiplo] = true;;
    }

    private void setPosicao(int i, int resultado){

        this.inicio = i*resultado;
        this.fim = (i+1) * resultado - 1;
    }

    static private void marcaZeroOuUm(){
            marcados[0] = true;
            marcados[1] = true;
    }

    static private Integer pegaProximoDivisor(int tamanhoVetor){
        for (Integer i = divisor; i.compareTo(tamanhoVetor) <= 0; i++) {
            if(!estaMarcado(i) && isNotDivisor(i)){
                return i;
            }
        }
        return tamanhoVetor;
    }

    static private boolean isNotDivisor(int num){
        return num != divisor;
    }

    private boolean isDivisivel(int num){
        return num % divisor == 0;
    }

    private void filtraMultiplos(){
        for (int i = this.inicio; i <= this.fim; i++) {
            if(!estaMarcado(i) && this.isDivisivel(i) && isNotDivisor(i))
                this.marcaNaoPrimo(i);
        }
    }

    static private boolean fimDoCrivo(int tam){
        return (divisor * divisor < tam);
    }

    // private void mostraSequencia(){
    //     System.out.print("thread indice ["+this.inicio+"] :");
    //     for (int i = this.inicio; i <= this.fim; i++) {
    //         if (estaMarcado(i)) {
    //             System.out.print("("+i+") ");
    //         }else{
    //             System.out.print(i+" ");
    //         }          
    //     }
    //     System.out.println();
    // }


    public static void main(String[] args)  throws InterruptedException {

        Scanner teclado = new Scanner(System.in);
		int numeroThreadsEscolhido;

        Integer resultado, tamanhoVetor;

        System.out.println("Entre com a quantidade de threads: ");
		numeroThreadsEscolhido = teclado.nextInt();

        System.out.println("Entre com o valor máximo: ");
		tamanhoVetor = teclado.nextInt();

        teclado.close();
        
        marcados = new boolean [tamanhoVetor + 1];

        marcaZeroOuUm();

        
        CrivoParaleloDados[] threads = new CrivoParaleloDados[numeroThreadsEscolhido];
        
        resultado = tamanhoVetor/numeroThreadsEscolhido;
       
        long tempo1, tempo2; 
       
        tempo1 = System.nanoTime(); 
        while (fimDoCrivo(tamanhoVetor)) {

            for (int i = 0; i < numeroThreadsEscolhido; i++) {
                threads[i] = new CrivoParaleloDados(i,resultado);
            }

            threads[numeroThreadsEscolhido-1].fim = tamanhoVetor;

            System.out.println("divisor: "+ divisor);
            //-----------Parte paralela-------------------
			for(int j = 0; j < numeroThreadsEscolhido; j++)
            threads[j].start();
            
            for(int j = 0; j < numeroThreadsEscolhido; j++)
            threads[j].join();
            //-----------Parte paralela-------------------
            // for(int j = 0; j < numeroThreadsEscolhido; j++)
            // threads[j].mostraSequencia();
            
            divisor = pegaProximoDivisor(tamanhoVetor); 
        }
        tempo2 = System.nanoTime(); 
        System.out.print("N = " + numeroThreadsEscolhido);
        System.out.println("\tTempo = " + String.valueOf((tempo2 - tempo1)/1000000) + " ms");
    }
}

File: crivoDeEratostenesComParalelismo/openMp/crivoParaleloDados.cpp
#include <iostream>
#include <omp.h>
#include <unordered_set>

using namespace std;


void setPosicao(int &inicio, int &fim, int i, int resultado){

    inicio = i*resultado;
    fim = (i+1) * resultado - 1;
}

void marcaZeroOuUm(std::unordered_set<int> &marcados ){
        marcados.emplace(0);
        marcados.emplace(1);
}

void marcaNaoPrimo(std::unordered_set<int> &marcados ,int multiplo){
        marcados.emplace(multiplo);
}

bool estaMarcado(std::unordered_set<int> marcados, int num){
    return marcados.count(num)>0;
}

bool isDivisivel(int num, int divisor){
    return num % divisor == 0;
}

bool isNotDivisor(int num, int divisor){
    return num != divisor;
}

void filtraMultipolos(std::unordered_set<int> &marcados,int num, int divisor){
    if(!estaMarcado(marcados,num) && isDivisivel(num,divisor) && isNotDivisor(num, divisor))
        marcaNaoPrimo(marcados,num);
}

bool fimDoCrivo(int divisor,int tam){
    return (divisor * divisor < tam);
}

int pegaProximoDivisor(int divisor,std::unordered_set<int> marcados,int tamanhoSequencia){
    for (int i = divisor; i <= (tamanhoSequencia); i++) {
        if(!estaMarcado(marcados,i) && isNotDivisor(i,divisor)){
            return i;
        }
    }
    return tamanhoSequencia;
}

void mostraSequencia(std::unordered_set<int> marcados, int tamanhoSequencia){
    for (int i = 0; i <= tamanhoSequencia; i++) {
        if (estaMarcado(marcados,i)) {
            cout<<"("<<i<<") ";
        }else{
            cout<<i<<" ";
        }          
    }
    cout<<endl;
}

int main (int argc, char *argv[])
{
    int qtdThreads, id, tamanhoSequencia;

    cout<<"Selecione o numero de threads:";
    cin>>qtdThreads;

    cout<<"Selecione tamanho da sequencia:";
    cin>>tamanhoSequencia;

    std::unordered_set<int> marcados = std::unordered_set<int>();

    double ti,tf;

    ti = omp_get_wtime();
    marcaZeroOuUm(marcados);
    int divisor = 2;
    int intervalo = tamanhoSequencia/qtdThreads; 


    while(fimDoCrivo(divisor,tamanhoSequencia))
    {
        omp_set_num_threads(qtdThreads);
        #pragma omp parallel for
            for (int i = 0 ; i <= tamanhoSequencia; i++)
            {
                filtraMultipolos(marcados,i,divisor);
            }   

        divisor = pegaProximoDivisor(divisor,marcados,tamanhoSequencia);
        std::cout<<"Divisor: "<<divisor<<std::endl;
    }
    tf = omp_get_wtime();
    //mostraSequencia(marcados,tamanhoSequencia);
    std::cout<<"Usando: "<<qtdThreads<<" threads"<<std::endl;
    std::cout<<"Tempo gasto: "<<tf-ti<<std::endl;
    return 0;
}
